File: rust_imp/src/f_ast/ast.rs
pub enum AST_Type {
    AST_VARIABLE_DEFINITION,
    AST_FUNCTION_DEFINITION,
    AST_VARIABLE,
    AST_FUNCTION_CALL,
    AST_STRING,
    AST_INT,
    AST_FLOAT,
    AST_BOOL,

    AST_COMPOUND,
    AST_LINE_COMMENT,
    AST_NOOP,
}

pub enum Variable_Type {
    VAR_STRING,
    VAR_INT,
    VAR_FLOAT,
    VAR_BOOL,
}

//I don't see how this structure will work in Rust. I'll actually have to start thinking for myself here.
pub struct AST {
    _type: AST_Type,

    line: usize,

    variable_definition_name: String,
    variable_definition_value: Box<AST>,

    function_definition_name: String,
    function_definition_body: Box<AST>,
    function_definition_args: Vec<AST>,
    function_definition_args_size: usize,

    variable_name: String,

    function_call_name: String,
    function_call_args: Vec<AST>,
    function_call_args_size: usize,

    compound_ast: Vec<AST>,
    compound_size: usize,
}

File: src/AST.c
#include "include/AST.h"


AST_T* init_ast(int type) {
    AST_T* ast = calloc(1, sizeof(struct AST_STRUCT));
    ast->type = type;
    ast->line = 0;

    ast->scope = (void*) 0;

    /* AST_VBARIABLE DEFITION */
    ast->variable_definition_variable_name = (void*) 0;
    ast->variable_definition_value = (void*) 0;

    /* AST FUNCTION DEFINITION */
    ast->function_definition_body = (void*) 0;
    ast->function_definition_name = (void*) 0;
    ast->function_def_args = (void*) 0;
    ast->function_def_args_size = 0;

    /* AST_VARIABLE */
    ast->variable_name = (void*) 0;
    
    /* AST_FUNCTION_CALL */
    ast->function_call_name = (void*) 0;
    ast->function_call_arguments = (void*) 0;
    ast->function_call_arguments_size = 0;

    /* AST_STRING */
    ast->string_value = (void*) 0;

    /* AST_COMPOUND */
    ast->compound_value = (void*)0;
    ast->compound_size = 0;

    return ast;
}

AST_T* init_ast_with_line(int type, unsigned int line) {
    AST_T* ast = calloc(1, sizeof(struct AST_STRUCT));
    ast->type = type;
    ast->line = line;

    ast->scope = (void*) 0;

    /* AST_VBARIABLE DEFITION */
    ast->variable_definition_variable_name = (void*) 0;
    ast->variable_definition_value = (void*) 0;

    /* AST FUNCTION DEFINITION */
    ast->function_definition_body = (void*) 0;
    ast->function_definition_name = (void*) 0;

    /* AST_VARIABLE */
    ast->variable_name = (void*) 0;
    
    /* AST_FUNCTION_CALL */
    ast->function_call_name = (void*) 0;
    ast->function_call_arguments = (void*) 0;
    ast->function_call_arguments_size = 0;

    /* AST_STRING */
    ast->string_value = (void*) 0;

    /* AST_COMPOUND */
    ast->compound_value = (void*)0;
    ast->compound_size = 0;

    return ast;
}